File: mobile/ios/Air/SubModules/Ledger/APDUHelpers.swift

import BleTransport
import Foundation


enum APDUHelpers {
    private static let LEDGER_SYSTEM: UInt8 = 0xB0
    static let LEDGER_CLA: UInt8 = 0xE0
    static let INS_VERSION: UInt8 = 0x03
    static let INS_ADDRESS: UInt8 = 0x05
    static let INS_SIGN_TX: UInt8 = 0x06
    static let INS_PROOF: UInt8 = 0x08
    static let INS_SIGN_DATA: UInt8 = 0x09
    
    static func currentApp() -> APDU {
        let bytes: [UInt8] = [
            LEDGER_SYSTEM,
            0x01,
            0x00,
            0x00,
            0x00
        ]
        return APDU(data: bytes)
    }
    
    static func getPublicKey(isTestnet: Bool, workChain: Int, index: Int) -> APDU {
        let paths = getLedgerAccountPath(isTestnet: isTestnet, workChain: Int32(workChain), index: Int32(index))
        var buffer: [UInt8] = [
            LEDGER_CLA,
            INS_ADDRESS,
            0x00,
            0x00,
            UInt8(1 + paths.count * 4)
        ]
        
        buffer.append(UInt8(paths.count))
        for path in paths {
            let hardenedPath = UInt32(bitPattern: path) + 0x80000000
            let bytes = withUnsafeBytes(of: hardenedPath.bigEndian) { Array($0) }
            buffer.append(contentsOf: bytes)
        }
        
        return APDU(data: buffer)
    }
    
    static func getLedgerAccountPath(isTestnet: Bool, workChain: Int32, index: Int32) -> [Int32] {
        [44, 607, isTestnet ? 1 : 0, workChain, index, 0]
    }
    
    static func decode(hex: String) -> [UInt8] {
        return stride(from: 0, to: hex.count, by: 2).map {
            let startIndex = hex.index(hex.startIndex, offsetBy: $0)
            let endIndex = hex.index(startIndex, offsetBy: 2)
            return UInt8(hex[startIndex..<endIndex], radix: 16) ?? 0
        }
    }
    
    static func decodeReadable(_ hex: String) -> [String] {
        let bytes = stride(from: 0, to: hex.count, by: 2).map {
            let startIndex = hex.index(hex.startIndex, offsetBy: $0)
            let endIndex = hex.index(startIndex, offsetBy: 2)
            return UInt8(hex[startIndex..<endIndex], radix: 16) ?? 0
        }
        
        var parts: [String] = []
        var current = ""
        
        for byte in bytes {
            let ch = Character(UnicodeScalar(byte))
            if ch.isASCII && ch >= " " && ch <= "~" {
                current.append(ch)
            } else {
                if !current.isEmpty {
                    parts.append(current)
                    current = ""
                }
            }
        }
        
        if !current.isEmpty {
            parts.append(current)
        }
        
        return parts
    }
}

File: mobile/android/air/SubModules/Ledger/src/main/java/org/mytonwallet/app_air/ledger/helpers/APDUHelpers.kt
package org.mytonwallet.app_air.ledger.helpers

import com.ledger.live.ble.extension.toHexString
import java.nio.ByteBuffer

class APDUHelpers {
    companion object {
        private val LEDGER_SYSTEM = 0xB0
        const val LEDGER_CLA = 0xE0
        const val INS_VERSION = 0x03
        const val INS_ADDRESS = 0x05
        const val INS_SIGN_TX = 0x06
        const val INS_PROOF = 0x08
        const val INS_SIGN_DATA = 0x09

        fun currentApp(): String {
            return byteArrayOf(
                LEDGER_SYSTEM.toByte(),
                0x01,
                0x00,
                0x00,
                0x00
            ).toHexString()
        }

        fun getLedgerAccountPath(isTestnet: Boolean, workChain: Int, index: Int): List<Int> {
            return listOf(44, 607, if (isTestnet) 1 else 0, workChain, index, 0)
        }

        fun getPublicKey(isTestnet: Boolean, workChain: Int, index: Int): String {
            val paths = getLedgerAccountPath(isTestnet, workChain, index)
            return (byteArrayOf(
                LEDGER_CLA.toByte(),
                INS_ADDRESS.toByte(),
                0x00,
                0x00,
                (1 + paths.size * 4).toByte()
            ) + ByteBuffer.allocate(1 + paths.size * 4).apply {
                put(paths.size.toByte())
                paths.map { (it + 0x80000000).toInt() }.forEach(this::putInt)
            }.array()).toHexString()
        }

        fun decode(hex: String): List<Int> {
            return hex.chunked(2).map { it.toInt(16) }
        }

        fun decodeReadable(hex: String): List<String> {
            val bytes = hex.chunked(2)
                .map { it.toInt(16).toByte() }
                .toByteArray()

            val parts = mutableListOf<String>()
            val current = StringBuilder()

            for (byte in bytes) {
                val ch = byte.toInt().toChar()
                if (ch in ' '..'~') {
                    current.append(ch)
                } else {
                    if (current.isNotEmpty()) {
                        parts.add(current.toString())
                        current.clear()
                    }
                }
            }

            if (current.isNotEmpty()) {
                parts.add(current.toString())
            }

            return parts
        }
    }
}
